modules/RTC/TrackStreamingStatus.js

```diff
@@ -0,0 +1,758 @@
+         *
+         * @type {number} amount of time in milliseconds
+         */
+        this.rtcMuteTimeout
+            = typeof options.rtcMuteTimeout === 'number'
+                ? options.rtcMuteTimeout : DEFAULT_RTC_MUTE_TIMEOUT;
+
+        /**
+         * This holds a timestamp indicating  when remote video track
+         * was RTC muted. The purpose of storing the timestamp is to
+         * avoid the transition to disconnected status in case of legitimate
+         * video mute operation where the signalling video muted event can
+         * arrive shortly after RTC muted event.
+         *
+         * The timestamp is measured in milliseconds obtained with
+         * <tt>Date.now()</tt>.
+         *
+         * FIXME merge this logic with NO_DATA_FROM_SOURCE event
+         *       implemented in JitsiLocalTrack by extending the event to
+         *       the remote track and allowing to set different timeout for
+         *       local and remote tracks.
+         *
+         * @type {number}
+         */
+        this.rtcMutedTimestamp = null;
+        logger.info(`RtcMuteTimeout set to: ${this.rtcMuteTimeout}`);
+
+        /**
+         * This holds the timestamp indicating when remote video track
+         * entered forwarded sources set. Track entering forwardedSources
+         *  will have connection status restoring and when we start receiving
+         * video will become active, but if video is not received for certain
+         * time {@link DEFAULT_RESTORING_TIMEOUT} that track connection status
+         * will become interrupted.
+         *
+         * @type {number}
+         */
+        this.enteredForwardedSourcesTimestamp = null;
+
+        /**
+         * This holds the restoring timeout callback ID
+         * scheduled using window.setTimeout.
+         *
+         * @type {number}
+         */
+        this.restoringTimer = null;
+
+        /**
+         * This holds the current connection status (along with all the internal events that happen
+         * while in that state).
+         *
+         * The goal is to send this information to the analytics backend for post-mortem analysis.
+         */
+        this.connectionStatusMap = {};
+    }
+
+    /**
+     * Gets the video frozen timeout for given source name.
+     * @return {number} how long are we going to wait since RTC video muted
+     * even, before a video track is considered frozen.
+     * @private
+     */
+    _getVideoFrozenTimeout() {
+        const sourceName = this.track.getSourceName();
+
+        return this.rtc.isInForwardedSources(sourceName)
+            ? this.rtcMuteTimeout
+            : this.conference.isP2PActive() ? this.p2pRtcMuteTimeout : this.outOfLastNTimeout;
+    }
+
+    /**
+     * Initializes <tt>TrackStreamingStatus</tt> and bind required event
+     * listeners.
+     */
+    init() {
+        // Handles P2P status changes
+        this._onP2PStatus = this.figureOutConnectionStatus.bind(this);
```

Comment:
Since `TrackStreamingStatus` is scoped to a single track, we no longer need the `ParticipantConnectionStatus#refreshConnectionStatusForAll` method. Each instances will now handle refreshing its own connection status.